### DISCUSSION TITLE
Is it possible to customise filename for submitted label?

### DISCUSSION THREAD

**Discussion by MalikTeng** [Project-MONAI/MONAILabel#1590] [Q&A] [answered]
I'm working with a 3D slicer and the monai-label plugin for heart MR image segmentation. I want to automatically keep track of the specific monai bundle used for labeling by adding its name as a suffix to the label file. Is this feasible, or do you have a better approach?

**Comment by diazandr3s** [comment]
Hi @MalikTeng,

This is a great question.

Are you starting MONAI Label pointing to a local folder? If yes, this is the class we use to manage the local datastore: https://github.com/Project-MONAI/MONAILabel/blob/main/monailabel/datastore/local.py#L86

That class creates and modifies the **datastore_V*.json** file we use to keep track of the volumes and labels.

These are the endpoints to modify the datastore: https://github.com/Project-MONAI/MONAILabel/blob/main/monailabel/endpoints/datastore.py#L93

When you submit the label using  Slicer, this is the method that is called: https://github.com/Project-MONAI/MONAILabel/blob/main/plugins/slicer/MONAILabel/MONAILabel.py#L1433

and this is the endpoint: https://github.com/Project-MONAI/MONAILabel/blob/main/monailabel/endpoints/datastore.py#L93

One idea would be to modify the _save_label_ method by adding more information to the submitted label name.

Hope this helps,

**Reply by MalikTeng** [reply]
Hi @diazandr3s,

Much appreciated your advice! It seems that I can address the issue by making some adjustments to specific sections of the code within those two scripts, which might need to clone and install the MONAI Label repository on my local machine. However, I couldn't locate precise instructions on how to do this correctly. 

Could you please guide me to the relevant documentation or resources that can assist me in this process?

Many thanks!

**Reply by diazandr3s** [reply]
Hi @MalikTeng,

Instructions for this sort of customization would be the code itself. 
Have you tried to modify the datastore to change the prediction name? Here is the part where this happens: https://github.com/Project-MONAI/MONAILabel/blob/main/monailabel/datastore/local.py#L496

**Comment by MalikTeng** [comment, marked as answer]
Hi @diazandr3s,

To enable bundle name tracking in the MONAI Label plugin with 3D Slicer, I followed these steps:

1. **Clone and Install MONAI Label**: Clone the MONAI Label repository to your local machine and install it in edit mode. For example, clone to `~/MONAILabel` and install using the command `pip install -e .`.

2. **Modify MONAILabel.py**: In the `MONAILabel.py` script, located at `~/Slicer-5.4.0-linux-amd64/slicer.org/Extensions-31938/MONAILabel/lib/Slicer-5.4/qt-scripted-modules`, alter the `self.logic.save_label` call. Add the bundle name used for segmentation by including `self.ui.segmentationModelSelector.currentText` to the `params` argument, forming a dictionary like `{"label_info": label_info, "model": model}`. Here, `model` represents the bundle name.

3. **Update local.py**: In the cloned MONAI Label repository, navigate to `~/MONAILabel/monailabel/datastore/local.py`. In function `save_label`, modify the `label_tag` in the logger info and label path lines to `label_info.get("model")`. This change ensures the bundle name passed in step 2 is used.

After these modifications, MONAI Label will save labels in folders named after the bundle's name. Note that this solution might vary based on different 3D Slicer versions and may require repeating the first step after updates to the MONAI Label plugin.

Looking forward to your feedback!

**Reply by diazandr3s** [reply]
Thanks for sharing this, @MalikTeng 
It makes sense.
